File: src/lib/store.js
import fs from "node:fs";

const databasePathname = new URL("../database.json", import.meta.url).pathname;

export async function getAdsByTerm(term) {
  const content = await fs.promises.readFile(databasePathname);

  const stored = JSON.parse(content.toString()).data;
  const storedTerm = stored.find((t) => t.term === term);

  if (!storedTerm) {
    return null;
  }

  return storedTerm.ads;
}

export async function saveAdsByTerm(term, ads) {
  const content = await fs.promises.readFile(databasePathname);

  const stored = JSON.parse(content.toString()).data;
  const storedTerm = stored.find((t) => t.term === term);

  if (storedTerm) {
    storedTerm.ads = ads;
    storedTerm.modifiedAt = Date.now();
  } else {
    stored.push({ term, ads, modifiedAt: Date.now() });
  }

  await fs.promises.writeFile(databasePathname, JSON.stringify(stored));
}

File: src/lib/store.py
from typing import Any
from entities.ad import Ad
import json


def get_ads_by_term(term: str):
    with open("./src/database.json", "r") as file:
        result: list[Ad] = []
        data = json.loads(file.read())["data"]

        for i in data:
            if i["term"] == term:
                for j in i["ads"]:
                    result.append(Ad.parse(j))

        file.close()
        return result


def save_ads_by_term(term: str, ads: list[Ad]):
    with open("./src/database.json", "wr") as file:
        data = json.loads(file.read())["data"]
        updated: list[dict[str, Any]] = []

        for i in data:
            if i["term"] != term:
                updated.append(i)

        updated.append({"term": term, "ads": ads})
        file.write(json.dumps({"data": updated}))
        file.close()
